printing/proof_diffs.ml

```diff
@@ -39,18 +39,32 @@ uses strikeout on removed text.
 
 open Pp_diff
 
+let term_color = ref true
+
+let write_color_enabled enabled =
+  term_color := enabled
+
+let color_enabled () = !term_color
+
```

Comment:
Umm, I general I am not fond of the diff code learning about state that belongs to the terminal layer. Indeed, I think the diffs code at this level should only have a job: perform diffs. It is up to the UI layers to decide how to call it / how to render the diffs.

For example, this seems bogus as a color-less UI could decide to render the diffs using symbols or some other approach. NACK on this part.

Reply:
Currently there's no way to render diffs in coqtop without color.  If diffs are turned on in coqtop when the terminal doesn't support color, you see the escape sequence printed.  Not a good result, so I wanted to prevent that.

Setting an option is a UI action, but AFAIK most options are not defined in UI-specific files.  

I didn't see any mechanism for the UI code to query command line arguments or environmental settings such as whether color is enabled, so I did what I did.

It seems to me the possibilities are:

1. Let the user turn on diffs anyway (i.e., remove the check)
2. Provide alternate code now (e.g. a better mechanism)
3. Keep it as it is, with the expectation that we refactor when # 2 becomes available later

What do you suggest?

Reply:
I think the terminal printing code should be fixed, that may involve transforming the "diffs on" check into "diffs on && term == good"

Reply:
That was my original idea, but I ran into two problems:
1. I need to know if I'm running coqtop alone or for CoqIDE.  If it's for CoqIDE, color is always available.  But I didn't see a clean way to determine that.  Is there a simple clean way to tell?
2. IIRC if init_color sets a variable in Coqtop to indicate that color is available (has command line arg+valid config), I get a circular linking error because Coqtop is in vernac while Proof_diffs is in printing.  I suppose I could get around that by adding a global variable in another file, but that's not too elegant.

Your thoughts?

Reply:
The point is that `Proof_diffs` should not care about colors: I think it has a single mission, to perform diff well, and return the diff object. It knows about printing, but not about what the clients are gonna do with the diff.

Then, the users [coqide, coqtop, serapi] should know whether they want to use the diff machinery or not.

Regarding color it is not so important, what you describe sounds like a bug of the terminal printing code which should gracefully fallback to a symbol-based printing. As this may be hard to implement, coqtop should indeed implement the diff logic consulting the color settings.

Reply:
Well, I can move some or all of the following code to Coqtop if that's what you mean (though I would prefer leaving the diffs option definition and the read/write in Proof_diffs)..  But any reference from Proof_diffs to Coqtop creates a circular dependency.  The way around that is to put the variables/functions in another file that both Proof_diffs and Coqtop can access.  The new file could be specific to this single problem or it could be a `Globals.ml` including various other bits of global data that present circularity problems.

> The point is that `Proof_diffs` should not care about colors: I think it has a single mission, to perform diff well, and return the diff object. It knows about printing, but not about what the clients are gonna do with the diff.

The diff code proper doesn't look at the color state and has no opinion about what clients will or may do with the result.  Only the Set/Unset Diffs commands--user interface--check for color support.

What do you suggest?

```
let term_color = ref true

let write_color_enabled enabled =
  term_color := enabled

let color_enabled () = !term_color

let diff_option = ref `OFF

let read_diffs_option () = match !diff_option with
| `OFF -> "off"
| `ON -> "on"
| `REMOVED -> "removed"

let write_diffs_option opt =
  let enable opt =
    if not (color_enabled ()) then
      CErrors.user_err Pp.(str "Enabling Diffs requires setting the \"-color\" command line argument to \"on\" or \"auto\".")
    else
      diff_option := opt
  in
  match opt with
  | "off" -> diff_option := `OFF
  | "on" -> enable `ON
  | "removed" -> enable `REMOVED
  | _ -> CErrors.user_err Pp.(str "Diffs option only accepts the following values: \"off\", \"on\", \"removed\".")

let _ =
  Goptions.(declare_string_option {
    optdepr = false;
    optname = "show diffs in proofs";
    optkey = ["Diffs"];
    optread = read_diffs_option;
    optwrite = write_diffs_option
  })


let show_diffs () = !diff_option <> `OFF;;
let show_removed () = !diff_option = `REMOVED;;
```

Reply:
> But any reference from Proof_diffs to Coqtop creates a circular dependency.

Precisely my point is that `Proof_diffs` should not learn anything about the console.